app/javascript/App/bundles/ExchangeListings/client.js

```diff
@@ -0,0 +1,88 @@
+import React, { Component, Fragment } from 'react'
+import debounce from 'debounce'
+import LayoutDesktop from '../../components/LayoutDesktop'
+import LayoutTablet from '../../components/LayoutTablet'
+import LayoutMobile from '../../components/LayoutMobile'
+import ListingsHeader from './components/ListingsHeader'
+import ListingsList from './components/ListingsList'
+import BodySection from './components/BodySection'
+
+class ExchangeListingsPage extends Component {
+  state = { showFilterPanel: false }
+
+  toggleFilterPanel() {
+    this.setState({
+      showFilterPanel: !this.state.showFilterPanel,
+    })
+  }
+
+  componentDidUpdate() {
+    window.addEventListener('resize', debounce(() => this.forceUpdate(), 500))
```

Comment:
This is a leak: new listener is being added on every re-render and doesn't get removed. Global subscriptions should be added only once in `componentDidMount` and removed on `componentWillUnmount`.

Reply:
@cryptoB0T event handler is not the same because it's the anonymous function. So it will register multiple times after every `componentDidUpdate`
https://jsfiddle.net/L9ahtn4k/1/
https://triangle717.wordpress.com/2015/12/14/js-avoid-duplicate-listeners/

Reply:
@cryptoB0T This is a leak in few ways. MDN doc states that `If multiple identical EventListeners ... the duplicate instances are discarded`, but `debounce` produces a brand new function on every update and it makes these listeners non-identical. To make them identical this function should be created only once and attached to a class instance, i.e.:

```js
updateOnResize = debounce(() => this.forceUpdate());

componentDidUpdate() {
  window.addEventListener('resize', this.updateOnResize);
}

// now `this.updateOnResize === this.updateOnResize` is always `true`
```

But even then there still is a leak: when component is unmounted from the DOM all added listeners are still around and stuck here forever since all instances required for removal are not available anymore and unmounting of a node from the DOM doesn't trigger auto-removal of all attached listeners.

So, instead of performing wasteful ops in `componentDidUpdate`, it's better to add one single listener in `componentDidMount` then on every `resize` this debounced handler will be triggered, plus to remove it in `componentWillUnmount`.